src/txn/txn_timestamp.c

```diff
@@ -902,46 +941,53 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
     __wt_config_init(session, &cparser, cfg[1]);
     while ((ret = __wt_config_next(&cparser, &ckey, &cval)) == 0) {
         WT_ASSERT(session, ckey.str != NULL);
-        if (WT_STRING_MATCH("commit_timestamp", ckey.str, ckey.len)) {
+        if (WT_STRING_MATCH("commit_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "commit", &commit_ts, &cval));
-            set_ts = true;
-        } else if (WT_STRING_MATCH("durable_timestamp", ckey.str, ckey.len)) {
+        else if (WT_STRING_MATCH("durable_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "durable", &durable_ts, &cval));
-            set_ts = true;
-        } else if (WT_STRING_MATCH("prepare_timestamp", ckey.str, ckey.len)) {
+        else if (WT_STRING_MATCH("prepare_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "prepare", &prepare_ts, &cval));
-            set_ts = true;
-        } else if (WT_STRING_MATCH("read_timestamp", ckey.str, ckey.len)) {
+        else if (WT_STRING_MATCH("read_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "read", &read_ts, &cval));
-            set_ts = true;
-        }
     }
     WT_RET_NOTFOUND_OK(ret);
 
-    /* Look for a commit timestamp. */
-    if (commit_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_commit_timestamp(session, commit_ts));
-
-    /*
-     * Look for a durable timestamp. Durable timestamp should be set only after setting the commit
-     * timestamp.
-     */
-    if (durable_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_durable_timestamp(session, durable_ts));
-    __wt_txn_publish_durable_timestamp(session);
+    return __wt_txn_set_timestamp_internal(session, commit_ts, durable_ts, prepare_ts, read_ts);
+}
 
-    /* Look for a read timestamp. */
-    if (read_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_read_timestamp(session, read_ts));
+/*
+ * __wt_txn_set_timestamp_numeric --
+ *     Directly set the commit timestamp in a transaction, bypassing parsing logic. Prefer this to
+ *     @ref __wt_txn_set_timestamp when string parsing is a performance bottleneck. The durable
+ *     timestamp must only be set after the commit timestamp.
+ */
+int
+__wt_txn_set_timestamp_numeric(WT_SESSION_IMPL *session, WT_TS_TXN_TYPE which, wt_timestamp_t ts)
+{
+    WT_DECL_RET;
 
-    /* Look for a prepare timestamp. */
-    if (prepare_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_prepare_timestamp(session, prepare_ts));
+    WT_TRET(__wt_txn_context_check(session, true));
```

Comment:
This should be `WT_RET()`, I think, isn't `__wt_txn_context_check()` a hard failure?

Reply:
I've double-checked and yes, I agree. Was the original `__wt_txn_set_timestamp` wrong here as well? I can't see why we wouldn't just bail on the `EINVAL` it could return.

Reply:
(Fixed my new function in c3075dc and left the original)

Reply:
It looks to me like `__wt_txn_set_timestamp()` is wrong as well.

Reply:
Thanks for checking, fixed in 09893e8